SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -122,51 +145,246 @@ - (void)process
+            if (oldReadReceipt && oldReadReceipt.timestamp > newReadReceipt.timestamp) {
+                // If there's an existing read receipt for the same thread with
+                // a newer timestamp, discard the new read receipt.
+                DDLogVerbose(@"%@ Ignoring redundant read receipt for linked devices.", self.tag);
+            } else {
+                DDLogVerbose(@"%@ Enqueuing read receipt for linked devices.", self.tag);
+                self.toLinkedDevicesReadReceiptMap[threadUniqueId] = newReadReceipt;
+            }
+
+            if ([self areReadReceiptsEnabled]) {
+                DDLogVerbose(@"%@ Enqueuing read receipt for sender.", self.tag);
+                NSMutableArray<NSNumber *> *_Nullable timestamps = self.toSenderReadReceiptMap[messageAuthorId];
+                if (!timestamps) {
+                    timestamps = [NSMutableArray new];
+                    self.toSenderReadReceiptMap[messageAuthorId] = timestamps;
+                }
+                [timestamps addObject:@(message.timestamp)];
+            }
 
-        NSString *messageAuthorId = message.messageAuthorId;
-        OWSAssert(messageAuthorId.length > 0);
+            [self scheduleProcessing];
+        }
+    });
+}
+
+#pragma mark - Read Receipts From Recipient
+
+- (void)processReadReceiptsFromRecipient:(OWSSignalServiceProtosReceiptMessage *)receiptMessage
+                                envelope:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    OWSAssert(receiptMessage);
+    OWSAssert(envelope);
+    OWSAssert(receiptMessage.type == OWSSignalServiceProtosReceiptMessageTypeRead);
+
+    if (![self areReadReceiptsEnabled]) {
+        DDLogInfo(@"%@ Ignoring incoming receipt message as read receipts are disabled.", self.tag);
+        return;
+    }
+
+    NSString *recipientId = envelope.source;
+    OWSAssert(recipientId.length > 0);
+
+    PBArray *timestamps = receiptMessage.timestamp;
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            for (int i = 0; i < timestamps.count; i++) {
+                UInt64 timestamp = [timestamps uint64AtIndex:i];
+
+                NSArray<TSOutgoingMessage *> *messages
+                    = (NSArray<TSOutgoingMessage *> *)[TSInteraction interactionsWithTimestamp:timestamp
+                                                                                       ofClass:[TSOutgoingMessage class]
+                                                                               withTransaction:transaction];
+                OWSAssert(messages.count <= 1);
+                if (messages.count > 0) {
+                    // TODO: We might also need to "mark as read by recipient" any older messages
+                    // from us in that thread.  Or maybe this state should hang on the thread?
```

Comment:
I've discussed with Moxie and I think this comment is obsolete, but I'll address it after the next PR is merged.  

We're not going to "imply previous messages were also read" for "sender/recipient" read receipts.